ledger/participant-integration-api/src/main/scala/platform/indexer/poc/BatchingParallelIngestionPipe.scala

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.indexer.poc
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.daml.ledger.participant.state.v1.{Offset, Update}
+import com.daml.platform.store.dao.JdbcLedgerDao
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+object BatchingParallelIngestionPipe {
+
+  def apply[IN, RUNBATCH](
+      submissionBatchSize: Long,
+      batchWithinMillis: Long,
+      inputMappingParallelism: Int,
+      inputMapper: Iterable[IN] => Future[RUNBATCH],
+      seqMapperZero: RUNBATCH,
+      seqMapper: (RUNBATCH, RUNBATCH) => RUNBATCH,
+      ingestingParallelism: Int,
+      ingester: RUNBATCH => Future[RUNBATCH],
+      tailer: (RUNBATCH, RUNBATCH) => RUNBATCH,
+      tailingRateLimitPerSecond: Int,
+      ingestTail: RUNBATCH => Future[RUNBATCH],
+      runStageUntil: Int,
+  ) // 1 - nothing   2 - input mapping   3 - seq mapping   4 - db ingestion   5 - tailing   6 - persist params
+  (source: Source[IN, NotUsed]): Source[Unit, NotUsed] = {
+    assert(runStageUntil >= 1 && runStageUntil <= 6)
+    val stage1 = source
+    val stage2 = stage1
+      .groupedWithin(
+        submissionBatchSize.toInt,
+        FiniteDuration(batchWithinMillis, "millis"),
+      ) // TODO .batch adds no latency to the pipe, but batch and mapAsync combination leads to dirac impulses in the forming batch sizes, which leads practically single threaded ingestion throughput at this stage.
```

Comment:
`..leads to dirac impulses.. ` It's been a while since I read this. Fancy analogy though